### PR TITLE
Tool for benchmarking read speed

### PR DESCRIPTION
**Pull request by ashish-goswami**
Fixes https://github.com/dgraph-io/badger/issues/801

---
This change is [Reviewable](https://reviewable.io/reviews/dgraph-io/badger/837)

### CODE REVIEW

**Comment by martinmr** [comment]
This is a draft PR but you requested reviews? Is this ready for review?

File: badger/cmd/bench.go [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cmd
+
+import "github.com/spf13/cobra"
```
Comment (mangalaman93) [inline badger/cmd/bench.go]: use brackets

File: badger/cmd/read_bench.go [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+	"context"
+	"fmt"
+	"math/rand"
+	"strings"
+	"sync/atomic"
+	"time"
+
+	"github.com/dgraph-io/badger"
+	"github.com/dgraph-io/badger/options"
+	"github.com/dgraph-io/badger/pb"
```
Comment (mangalaman93) [inline badger/cmd/read_bench.go]: separate badger imports from other imports

File: badger/cmd/read_bench.go [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+
+var readBenchCmd = &cobra.Command{
+	Use:   "read",
+	Short: "Read data from Badger randomly to benchmark read speed.",
+	Long: `
+This command reads data from existing Badger database randomly using multiple go routines.`,
+	RunE: readBench,
+}
+
+var sizeRead, entriesRead uint64
```
Comment (mangalaman93) [inline badger/cmd/read_bench.go]: use `var ()` format

File: badger/cmd/read_bench.go [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+		&loadingMode, "loading-mode", "mmap", "Mode for accessing SSTables and value log files. "+
+			"Valid loading modes are fileio and mmap.")
+}
+
+func readBench(cmd *cobra.Command, args []string) error {
+	rand.Seed(time.Now().Unix())
+
+	dur, err := time.ParseDuration(duration)
+	if err != nil {
+		return y.Wrapf(err, "unable to parse duration")
```
Comment (mangalaman93) [inline badger/cmd/read_bench.go]: Use `errors.Errorf` instead, everywhere!

File: badger/cmd/read_bench.go [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+		return nil
+	}
+
+	fmt.Println("*********************************************************")
+	fmt.Println("Starting to benchmark Reads")
+	fmt.Println("*********************************************************")
+	c := y.NewCloser(0)
+	startTime = time.Now()
+	for i := 0; i < numGoroutines; i++ {
+		c.AddRunning(1)
```
Comment (mangalaman93) [inline badger/cmd/read_bench.go]: can just do c.AddRunning(numGoroutines) once

File: badger/cmd/read_bench.go [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+			key := keys[r.Int31n(int32(len(keys)))]
+			atomic.AddUint64(&sizeRead, lookupForKey(db, key))
+			atomic.AddUint64(&entriesRead, 1)
+		}
+	}
+}
+
+func lookupForKey(db *badger.DB, key []byte) (sz uint64) {
+	err := db.View(func(txn *badger.Txn) error {
+		itm, err := txn.Get(key)
```
Comment (mangalaman93) [inline badger/cmd/read_bench.go]: should avoid y.Check* functions

File: badger/cmd/read_bench.go [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+	count := 0
+	stream := db.NewStream()
+
+	// overide stream.KeyToList as we only want keys. Also
+	// we can take only first version for the key.
+	stream.KeyToList = func(key []byte, itr *badger.Iterator) (*pb.KVList, error) {
+		list := &pb.KVList{}
+		// Since stream framework copies the item's key while calling
+		// KeyToList, we can directly append key to list.
+		list.Kv = append(list.Kv, &pb.KV{
```
Comment (mangalaman93) [inline badger/cmd/read_bench.go]: can do this in single line

File: badger/cmd/read_bench.go [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+// getSampleKeys uses stream framework internally, to get keys in random order.
+func getSampleKeys(db *badger.DB) ([][]byte, error) {
+	var keys [][]byte
+	count := 0
+	stream := db.NewStream()
+
+	// overide stream.KeyToList as we only want keys. Also
+	// we can take only first version for the key.
+	stream.KeyToList = func(key []byte, itr *badger.Iterator) (*pb.KVList, error) {
+		list := &pb.KVList{}
```
Comment (mangalaman93) [inline badger/cmd/read_bench.go]: list is a package name, can avoid it as a variable name

File: badger/cmd/read_bench.go (hunk 6 above)
Comment (mangalaman93) [reply badger/cmd/read_bench.go]: should use `panic` instead.

**Pull request merged**